src/plugins/graph/adornments/plotted-function/plotted-variables/plotted-variables-adornment-model.ts

```diff
@@ -0,0 +1,108 @@
+import { uniqueId } from "../../../../../utilities/js-utils";
+import { SharedVariables, SharedVariablesType } from "../../../../shared-variables/shared-variables";
+import { IAdornmentModel } from "../../adornment-models";
+import { kPlottedVariablesType } from "./plotted-variables-adornment-types";
+import { PlottedFunctionAdornmentModel } from "../plotted-function-adornment-model";
+
+function getSharedVariables(node: IAnyStateTreeNode) {
+  const sharedModelManager = getSharedModelManager(node);
+  if (sharedModelManager?.isReady) {
+    const tile = getTileModel(node);
+    if (tile) {
+      const sharedVariables = sharedModelManager.findFirstSharedModelByType(SharedVariables, tile.id);
+      if (sharedVariables) return sharedVariables as SharedVariablesType;
+    }
+  }
+}
+
+export const PlottedVariablesInstance = types.model("PlottedVariablesInstance", {})
+  .props({
+    xVariableId: types.maybe(types.string),
+    yVariableId: types.maybe(types.string)
+  })
+  .volatile(self => ({
+    variablesCopy: undefined as VariableType[] | undefined,
+    xVariableCopy: undefined as VariableType | undefined,
+    yVariableCopy: undefined as VariableType | undefined
+  }))
+  .views(self => ({
+    get sharedVariables() {
+      return getSharedVariables(self);
+    }
+  }))
+  .views(self => ({
+    get xVariable() {
+      return self.sharedVariables?.variables.find(variable => variable.id === self.xVariableId);
+    },
+    get yVariable() {
+      return self.sharedVariables?.variables.find(variable => variable.id === self.yVariableId);
+    }
+  }))
+  .actions(self => ({
+    computeY(x: number) {
+      if (self.variablesCopy && self.xVariableCopy && self.yVariableCopy) {
+        self.xVariableCopy.setValue(x);
+        const dependentValue = self.yVariableCopy.computedValue;
+        return dependentValue ?? NaN;
+      }
+      return NaN;
+    },
+    disposeCompute() {
+      self.xVariableCopy = undefined;
+      self.yVariableCopy = undefined;
+      if (self.variablesCopy) destroy(self.variablesCopy);
+      self.variablesCopy = undefined;
+    },
+    setXVariableId(variableId?: string) {
+      self.xVariableId = variableId;
+    },
+    setYVariableId(variableId?: string) {
+      self.yVariableId = variableId;
+    }
+  }))
+  .actions(self => ({
+    setupCompute() {
+      self.variablesCopy = types.array(Variable).create(
+        self.sharedVariables ? getSnapshot(self.sharedVariables.variables) : []
+      );
+      self.xVariableCopy = self.variablesCopy?.find(variable => variable.id === self.xVariableId);
+      self.yVariableCopy = self.variablesCopy?.find(variable => variable.id === self.yVariableId);
+      return { computeY: self.computeY, dispose: self.disposeCompute };
+    }
+  }));
+
+export const PlottedVariablesAdornmentModel = PlottedFunctionAdornmentModel
+  .named("PlottedVariablesAdornmentModel")
+  .props({
+    plottedVariables: types.map(PlottedVariablesInstance),
```

Comment:
Is Map in MST stable in its iteration order?  It would be good to have the order shown in the legend be predictable. (but we can deal with this later when we implement multiple traces)

Reply:
I'm not sure about this one. I believe the order should be consistent, at least within a single session. The order might change when you reload the page.